src/Core/Event.js

```diff
@@ -0,0 +1,45 @@
+/**
+ * @type {Object}
+ */
+var Event = {
+	/**
+	 * @param {Element} element
+	 * @param {string} type
+	 */
+	addEventListener: function addEventListener (element, type) {
+		if (!this[type]) {
+			this[type] = new Map()
+			DOMEvent(type.substring(2), this, true)
+		}
+
+		this[type].set(DOMNode(element), element)
+	},
+	/**
+	 * @param {Event} event
+	 */
+	handleEvent: function handleEvent (event) {
+		var type = 'on'+event.type
+		var element = this[type].get(event.target)
```

Comment:
should be `event.currentTarget` I think

Reply:
This uses event delegation as an optimization, so `currentTarget` would point to `document`.

Reply:
interesting, so this means that https://github.com/thysultan/dio.js/issues/24 is impossible to fix

Reply:
...While on the other hand `e.target` will point to what we expected `currentTarget` to point to before. The situation is inverted but we still have all the references we need to implement what you suggested in #24.

Reply:
Where does event propagation occur? I don't understand how events bubble up the tree

Reply:
Bubbling is handled by the browser, we add a single event here as an `EventListener` interface to the root document Node src/DOM/DOM.js#30 DOMEvent.

Reply:
Not the DOM tree, bubbling up the component tree
e.g.
```
<div onclick=alert(1)>
  <button>
</div>
```
currently the delegator will check `e.target` which is `<button>` and has no listeners (and not call any callback); when in fact it should propagate up the tree and trigger the event on `<div>`

Reply:
With event delegation it is inverted, so `e.target` will be `<div>` and `e.currentTarget` will be `document`, at which point all that's left is to try to retrieve the corresponding element `src/Event.js/handleEvent#22` that should have been previously set at `src/Event.js/addEventListener#15`

Reply:
So there's no bubbling in the component space, it happens in O(1) time and i think because we set the useCapture param in [Node.addEventListener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) it should also be O(1) time in the DOM space.

Reply:
I see, yes you right, we should handle that.

Reply:
Thanks for pointing this out, updated the implementation.